zkevm-circuits/src/mpt_circuit/storage_leaf.rs

```diff
@@ -0,0 +1,436 @@
+                &mut cb.base,
+                config.main_data.proof_type.expr(),
+                ProofType::StorageChanged.expr(),
+            );
+            config.is_non_existing_storage_proof = IsEqualGadget::construct(
+                &mut cb.base,
+                config.main_data.proof_type.expr(),
+                ProofType::StorageDoesNotExist.expr(),
+            );
+
+            // Drifted leaf handling
+            config.drifted = DriftedGadget::construct(
+                cb,
+                &config.parent_data,
+                &config.key_data,
+                &key_rlc,
+                &value_rlp_rlc,
+                &drifted_item,
+                &ctx.r,
+            );
+
+            // Wrong leaf handling
+            config.wrong = WrongGadget::construct(
+                cb,
+                a!(ctx.mpt_table.key_rlc),
+                config.is_non_existing_storage_proof.expr(),
+                &config.rlp_key[true.idx()].key_value,
+                &key_rlc[true.idx()],
+                &wrong_item,
+                config.is_in_empty_trie[true.idx()].expr(),
+                config.key_data[true.idx()].clone(),
+                &ctx.r,
+            );
+
+            // For non-existing proofs the tree needs to remain the same
+            ifx! {config.is_non_existing_storage_proof => {
+                require!(config.main_data.root => config.main_data.root_prev);
+                require!(key_rlc[true.idx()] => key_rlc[false.idx()]);
+            }}
+
+            // Put the data in the lookup table
+            let proof_type = matchx! {
+                config.is_storage_mod_proof => ProofType::StorageChanged.expr(),
+                config.is_non_existing_storage_proof => ProofType::StorageDoesNotExist.expr(),
+                _ => ProofType::Disabled.expr(),
+            };
+            let key_rlc = ifx! {config.is_non_existing_storage_proof => {
+                a!(ctx.mpt_table.key_rlc)
+            } elsex {
+                key_rlc[false.idx()].expr()
+            }};
+            ctx.mpt_table.constrain(
+                meta,
+                &mut cb.base,
+                config.main_data.address_rlc.expr(),
+                proof_type,
+                key_rlc,
+                value_rlc[true.idx()].expr(),
+                value_rlc[false.idx()].expr(),
+                config.main_data.root_prev.expr(),
+                config.main_data.root.expr(),
+            );
+        });
+
+        config
+    }
+
+    pub fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        ctx: &MPTConfig<F>,
+        pv: &mut MPTState<F>,
+        offset: usize,
+        node: &Node,
+        rlp_values: &[RLPItemWitness],
+    ) -> Result<(), Error> {
+        let storage = &node.storage.clone().unwrap();
+
+        let key_items = [
+            rlp_values[StorageRowType::KeyS as usize].clone(),
+            rlp_values[StorageRowType::KeyC as usize].clone(),
+        ];
+        let value_item = [
```

Comment:
@Brecht: It seems now the RLP bytes length for storage value is always 1. This actually never happens - because it's either 0 (when value is of length 1) or 2 (when value is of length bigger than 1). None of the constraints fail because it affects only the RLC of the value, which gets stored into the MPT table and currently we have no lookups into it yet.

Shall I modify this?

Reply:
I had the storage values wrong for a long while so it wouldn't surprise me if I still have it wrong! 

Seems like I at least documented what I think should happen above:

>  // Because the storage value is an rlp encoded string inside another rlp encoded
// string (leaves are always encoded as [key, value], with
// `value` here containing a single stored value) the stored
// value is either stored directly in the RLP encoded string if short, or stored
// wrapped inside another RLP encoded string if long.

So the RLP value stored here is the internal value, not the main RLP value (see line 114 for that one). So this one will only be used when the RLP value is long. I hope this makes sense.

Reply:
Ah, true! The value was correctly computed!

However, I find a bit confusing that in the case of the short RLP, the descriptor contains something like `value_rlp_bytes":[[1],[161]]`. In this example, the previous value RLP is short and the current value RLP is long. Thus, in the short case, `value_rlp_bytes` stores the actual value (being `1` in this case) and not the RLP byte, while the `value` is `[0,0,...]`.

In https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1353/commits/92ef471f51e5c70b7e3ececa7a2a3b3b34d172fb I changed to `value_rlp_bytes":[[],[161]]`
and `value` to be `[1,0,...]`. Can you have a look if it seems right to you?

Reply:
I think this could work, but I also think it complicates things a bit as well because it breaks the meaning of what "short" is. `value_rlp_bytes":[[1],[161]]` for me makes sense (well, RLP encoding itself is a bit crazy) because the `1` is indeed the RLP byte, it also just happens to be the value.

So moving the `1` to the value bytes normally used for long values I think works, but also a bit non-standard compared to how short values are handled elsewhere (`RLPValueGadget` functions currently would return the wrong values for most things, though I guess this would be fixable by having a special mode for no data, instead of reusing the short flag).

Maybe another argument is that the storage leaf code works similarly to how the account leaf code works before, it's just that the account leaf always stores multiple values so the values stored in `value_rlp_bytes` are always actual RLP bytes.

But I don't really have a strong opinion on it.

Reply:
My impression was that `value_rlp_bytes` is used for RLP bytes (the ones that specify the length, not the actual value). But I don't have a strong opinion either and thus I reverted to the previous (your) version. Thanks, Brecht!